### DISCUSSION TITLE
Inherit belongsTo() relations form parent Model

### DISCUSSION THREAD

**Discussion by ad-on-is** [adonisjs/core#3581] [Help]
Many of my models share the same properties/relations. I'd want to extract these in a separate parent class

While columns such as `createdAt` work just fine, the relations throw an error, in this case the `belongsTo()`

```ts
export default class BaseAppModel extends compose(BaseModel, SoftDeletes) {
  @column.dateTime({ autoCreate: true })
  public createdAt: DateTime

  @column.dateTime({ autoCreate: true, autoUpdate: true })
  public updatedAt: DateTime

  @column()
  public creatorId: string

  @belongsTo(() => User, {
    foreignKey: 'creatorId',
  })
  public creator: BelongsTo<typeof User>


}
```

```ts
export default class User extends BaseAppModel {}
```

When I do something like ` return User.query()` in my Controller, I get the following error.

```json
{
	"message": "Cannot read properties of undefined (reading 'primaryKey')",
	"stack": "TypeError: Cannot read properties of undefined (reading 'primaryKey')\n    at SnakeCaseNamingStrategy.relationLocalKey (/home/adonis/Development/Node/my_cool_project/node_modules/.pnpm/@adonisjs+lucid@16.3.2_@adonisjs+core@5.4.2+pg@8.7.1/node_modules/@adonisjs/lucid/build/src/Orm/NamingStrategies/SnakeCase.js:40:33)\n    at BelongsTo.boot (/home/adonis/Development/Node/my_cool_project/node_modules/.pnpm/@adonisjs+lucid@16.3.2_@adonisjs+core@5.4.2+pg@8.7.1/node_modules/@adonisjs/lucid/build/src/Orm/Relations/BelongsTo/index.js:144:47)\n    at /home/adonis/Development/Node/my_cool_project/node_modules/.pnpm/@adonisjs+lucid@16.3.2_@adonisjs+core@5.4.2+pg@8.7.1/node_modules/@adonisjs/lucid/build/src/Orm/BaseModel/index.js:504:31\n    at Map.forEach (<anonymous>)\n    at /home/adonis/Development/Node/my_cool_project/node_modules/.pnpm/@adonisjs+lucid@16.3.2_@adonisjs+core@5.4.2+pg@8.7.1/node_modules/@adonisjs/lucid/build/src/Orm/BaseModel/index.js:502:19\n    at defineStaticProperty (/home/adonis/Development/Node/my_cool_project/node_modules/.pnpm/@poppinss+utils@3.3.1/node_modules/@poppinss/utils/build/src/defineStaticProperty.js:50:53)\n    at Function.$defineProperty (/home/adonis/Development/Node/my_cool_project/node_modules/.pnpm/@adonisjs+lucid@16.3.2_@adonisjs+core@5.4.2+pg@8.7.1/node_modules/@adonisjs/lucid/build/src/Orm/BaseModel/index.js:433:42)\n    at Function.boot (/home/adonis/Development/Node/my_cool_project/node_modules/.pnpm/@adonisjs+lucid@16.3.2_@adonisjs+core@5.4.2+pg@8.7.1/node_modules/@adonisjs/lucid/build/src/Orm/BaseModel/index.js:500:14)\n    at decorateAsColumn (/home/adonis/Development/Node/my_cool_project/node_modules/.pnpm/@adonisjs+lucid@16.3.2_@adonisjs+core@5.4.2+pg@8.7.1/node_modules/@adonisjs/lucid/build/src/Orm/Decorators/index.js:23:15)\n    at DecorateProperty (/home/adonis/Development/Node/my_cool_project/node_modules/.pnpm/reflect-metadata@0.1.13/node_modules/reflect-metadata/Reflect.js:553:33)"
}
```

Am I doing something wrong, or is this not possible?

**Comment by omidamraei** [comment]
I'm also facing the same problem.

**Comment by thetutlage** [comment]
What happens when you call `User.boot()` inside the User model?

```
export default class User extends BaseAppModel {}

User.boot()
```

**Reply by omidamraei** [reply]
@thetutlage Nothing for me. The same error occurs even If I extend any other model (rather than User) from BaseAppModel.

```ts
function ShareableModel<T extends NormalizeConstructor<LucidModel>>(superclass: T) {
  class AShareableModel extends superclass {
    @column({ isPrimary: true })
    public id: number;

    @column()
    public ownerId: number;

    @belongsTo(() => User, {
      foreignKey: 'ownerId'
    })
    public owner: BelongsTo<typeof User>;

  }

  return AShareableModel;
}

class Project extends compose(BaseModel, ShareableModel) {}
```

Then calling `User.find(1)` throws the same error:

![image](https://user-images.githubusercontent.com/3328211/164943547-0acdfed3-a280-424e-b26b-3d6dc8597d04.png)

It worth to mention this error occurs only if the mixin or parent class has a belongsTo relation. For other type of relations, it works.

**Reply by thetutlage** [reply]
Can you please create a fresh project, reproduce this error and share it with me? Because I am not able to do so

**Reply by omidamraei** [reply]
@thetutlage Here I attached a fresh project which reproduce the error.
[reproduction.zip](https://github.com/adonisjs/core/files/8582568/reproduction.zip)

There is a mixin:
- Shareable (implements a belongsTo relation to User using ownerId)

There are 3 models:
- User (which has many Projects)
- Project (which has many Items)
- Item (which extends Shareable)

Just by running `npm run dev`, the error occurs.

**Reply by omidamraei** [reply]
@thetutlage No chance?

**Comment by Melchyore** [comment]
I am also facing the same problem. I have tried many approaches but unfortunately without success.

**Comment by Melchyore** [comment]
After some investigation, it seems like the decorator isn't registering the relationship correctly. The first argument of the decorator which should return the related model, returns `undefined` actually.

Maybe it should be converted into an issue in the @adonisjs/lucid repository? I can provide a repo to demonstrate the issue, if needed.

**Comment by isimisi** [comment]
Anyone found a fix? - it seems like the child models' boot method never gets called?

**Reply by IArnaut2** [reply]
NOPE!
